src/main/java/org/asciidoctor/maven/process/SourceDocumentFinder.java

```diff
@@ -0,0 +1,75 @@
+package org.asciidoctor.maven.process;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.List;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Finds all source documents inside a source directory.
+ * It traverses the source directory and all subdirectories. It can match custom file extensions.
+ * If none are specified, it matches <code>.asc</code>, <code>.asciidoc</code>, <code>.ad</code> and <code>.adoc</code>.
+ * Files and directories starting with an underscore (<code>_</code>) are ignored.
+ *
+ * @author stdll
+ */
+public class SourceDocumentFinder {
+
+    /** Pattern for matching standard file extensions. */
+    private static final String STANDARD_FILE_EXTENSIONS_PATTERN = "^[^_.].*\\.a((sc(iidoc)?)|d(oc)?)$";
+
+    /** Prefix for matching custom file extensions. */
+    private static final String CUSTOM_FILE_EXTENSIONS_PATTERN_PREFIX = "^[^_.].*\\.(";
+
+    /** Suffix for matching custom file extensions. */
+    private static final String CUSTOM_FILE_EXTENSIONS_PATTERN_SUFFIX = ")$";
+
+    /**
+     * Finds all source documents inside the source directory with standard file extensions.
+     *
+     * @param sourceDirectory source directory
+     * @return the list of all matching source documents.
+     */
+    public List<File> find(Path sourceDirectory) {
+        return find(sourceDirectory, Pattern.compile(STANDARD_FILE_EXTENSIONS_PATTERN));
+    }
+
+    /**
+     * Finds all source documents inside the source directory with custom file extensions.
+     *
+     * @param sourceDirectory          source directory
+     * @param sourceDocumentExtensions custom file extensions
+     * @return the list of all matching source documents.
+     */
+    public List<File> find(Path sourceDirectory, List<String> sourceDocumentExtensions) {
+        String extensionPattern = sourceDocumentExtensions.stream()
+            .reduce(CUSTOM_FILE_EXTENSIONS_PATTERN_PREFIX, (result, s) -> result.concat("|").concat(s))
```

Comment:
I suspect this adds `|` at the beginning of each file extension, so the pattern ends looking like `...(|pattern1|pattern2|pattern3)...` which means files with empty extension (`file.`) should be returned.

We could try fixing with `for + java.util.StringJoiner` or joining collector :thinking: 
```suggestion
String extensionPattern = sourceDocumentExtensions.stream()
                .collect(Collectors.joining("|", CUSTOM_FILE_EXTENSIONS_PATTERN_PREFIX, CUSTOM_FILE_EXTENSIONS_PATTERN_SUFFIX));
```